Missing trailing expression in block with `let else` yields incorrectly spanned error message

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#![feature(let_else)]

pub fn test(a: Option<u32>) -> Option<u32> {
    let Some(_) = a else { return None; };
    println!("Foo");
}

fn main() {}
```

I expected to see this happen: A helpful error message pointing to the end of the block.

Instead, this happened: An error message with a wrong span:

```rust
error[E0308]: mismatched types
 --> test.rs:4:5
  |
4 |     let Some(_) = a else { return None; };
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `Option`, found `()`
  |
  = note:   expected enum `Option<u32>`
          found unit type `()`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (b17226fcc 2022-02-18)
binary: rustc
commit-hash: b17226fcc11587fed612631be372a5b4cb89988a
commit-date: 2022-02-18
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 14.0.0
```

cc #87335

@rustbot labels: +T-compiler +F-let-else +requires-nightly +A-diagnostics

FTR this is the error for the `if let` equivalent:

```Rust
pub fn test(a: Option<u32>) -> Option<u32> {
    if let Some(_) = a {
    } else {
        return None;
    }
    println!("Foo");
}
```

```
error[E0308]: mismatched types
 --> src/main.rs:3:32
  |
3 | pub fn test(a: Option<u32>) -> Option<u32> {
  |        ----                    ^^^^^^^^^^^ expected enum `Option`, found `()`
  |        |
  |        implicitly returns `()` as its body has no tail or `return` expression
  |
  = note:   expected enum `Option<u32>`
          found unit type `()`
```

Hmm if you move the println into the `if let`, it gives a similar error to the one emitted by `let else`:

```Rust
pub fn test(a: Option<u32>) -> Option<u32> {
    if let Some(_) = a {
        println!("Foo");
    } else {
        return None;
    }
}
```
gives
```
error[E0308]: mismatched types
 --> src/main.rs:2:24
  |
2 |       if let Some(_) = a {
  |  ________________________^
3 | |         println!("Foo");
4 | |     } else {
  | |_____^ expected enum `Option`, found `()`
  |
  = note:   expected enum `Option<u32>`
          found unit type `()`
```

--- a/src/test/ui/let-else/issue-94176.rs
+++ b/src/test/ui/let-else/issue-94176.rs
@@ -1,0 +1,10 @@
+// Issue #94176: wrong span for the error message of a mismatched type error,
+// if the function uses a `let else` construct.
+#![feature(let_else)]
+
+pub fn test(a: Option<u32>) -> Option<u32> { //~ ERROR mismatched types
+    let Some(_) = a else { return None; };
+    println!("Foo");
+}
+
+fn main() {}

--- a/src/test/ui/let-else/issue-94176.stderr
+++ b/src/test/ui/let-else/issue-94176.stderr
@@ -1,0 +1,19 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-94176.rs:5:32
+   |
+LL | pub fn test(a: Option<u32>) -> Option<u32> {
+   |        ----                    ^^^^^^^^^^^ expected enum `Option`, found `()`
+   |        |
+   |        implicitly returns `()` as its body has no tail or `return` expression
+   |
+   = note:   expected enum `Option<u32>`
+           found unit type `()`
+help: consider returning the local binding `a`
+   |
+LL ~     println!("Foo");
+LL +     a
+   |
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/let-else/let-else-then-diverge.rs
+++ b/src/test/ui/let-else/let-else-then-diverge.rs
@@ -1,0 +1,19 @@
+//
+// popped up in in #94012, where an alternative desugaring was
+// causing unreachable code errors
+
+#![feature(let_else)]
+#![deny(unused_variables)]
+#![deny(unreachable_code)]
+
+fn let_else_diverge() -> bool {
+    let Some(_) = Some("test") else {
+        let x = 5; //~ ERROR unused variable: `x`
+        return false;
+    };
+    return true;
+}
+
+fn main() {
+    let_else_diverge();
+}

--- a/src/test/ui/let-else/let-else-then-diverge.stderr
+++ b/src/test/ui/let-else/let-else-then-diverge.stderr
@@ -1,0 +1,14 @@
+error: unused variable: `x`
+  --> $DIR/let-else-then-diverge.rs:11:13
+   |
+LL |         let x = 5;
+   |             ^ help: if this is intentional, prefix it with an underscore: `_x`
+   |
+note: the lint level is defined here
+  --> $DIR/let-else-then-diverge.rs:6:9
+   |
+LL | #![deny(unused_variables)]
+   |         ^^^^^^^^^^^^^^^^
+
+error: aborting due to previous error
+

